cmd/encryption-v1.go

```diff
@@ -230,50 +124,27 @@ func ParseSSECustomerHeader(header http.Header) (key []byte, err error) {
 		// will always fail -> r.TLS is always nil even for TLS requests.
 		return nil, errInsecureSSERequest
 	}
-	if algorithm := header.Get(SSECustomerAlgorithm); algorithm != SSECustomerAlgorithmAES256 {
-		return nil, errInvalidSSEAlgorithm
-	}
-	if header.Get(SSECustomerKey) == "" {
-		return nil, errMissingSSEKey
-	}
-	if header.Get(SSECustomerKeyMD5) == "" {
-		return nil, errMissingSSEKeyMD5
+	if crypto.S3.IsRequested(header) && crypto.SSEC.IsRequested(header) {
+		return key, crypto.ErrIncompatibleEncryptionMethod
 	}
 
-	key, err = base64.StdEncoding.DecodeString(header.Get(SSECustomerKey))
-	if err != nil {
-		return nil, errInvalidSSEKey
-	}
-
-	if len(key) != SSECustomerKeySize {
-		return nil, errInvalidSSEKey
-	}
-	// Make sure we purged the keys from http headers by now.
-	header.Del(SSECustomerKey)
-
-	keyMD5, err := base64.StdEncoding.DecodeString(header.Get(SSECustomerKeyMD5))
-	if err != nil {
-		return nil, errSSEKeyMD5Mismatch
-	}
-	if md5Sum := md5.Sum(key); !bytes.Equal(md5Sum[:], keyMD5) {
-		return nil, errSSEKeyMD5Mismatch
-	}
-	return key, nil
+	k, err := crypto.SSEC.ParseHTTP(header)
+	return k[:], err
 }
 
 // This function rotates old to new key.
 func rotateKey(oldKey []byte, newKey []byte, bucket, object string, metadata map[string]string) error {
-	delete(metadata, SSECustomerKey) // make sure we do not save the key by accident
+	delete(metadata, crypto.SSECKey) // make sure we do not save the key by accident
```

Comment:
I really discourage defining metadata names in crypto package. Someone can change the value of crypto.SSEKey (or similar metadata names) and run crypto tests that will still be passing.

Reply:
@vadmeste I don't think so. See, `crypto` is e.g. responsible for parsing SSE metadata.
1. Based on our wrong package layout it is not possible to define metadata entry keys in `cmd` and use in `crypto`
2. Even if you could what prevents someone from changing the headers in `cmd` and not running the `crypto` tests?

That is something for CI and reviewers, changing the format is a breaking change - and requires a whole approval cycle...

Defining metadata entry keys `crypto` has several adv. like group and separate responsibility, (better) testable code, documentation, a.s.o.

Reply:
agree with @aead on this, would make the code more testable. Currently, the code was duplicated a lot in encryption.go and is more error prone.